relay-server/src/actors/processor.rs

```diff
@@ -1494,11 +1514,158 @@ impl EnvelopeProcessorService {
+                event_id: None,
+                remote_addr: None,
+                category: DataCategory::Transaction,
+                quantity: source_quantities.transactions as u32,
+            });
+        }
+        if source_quantities.profiles > 0 {
+            self.inner.outcome_aggregator.send(TrackOutcome {
+                timestamp,
+                scoping,
+                outcome: Outcome::RateLimited(reason_code.clone()),
+                event_id: None,
+                remote_addr: None,
+                category: DataCategory::Profile,
+                quantity: source_quantities.profiles as u32,
+            });
+        }
+
+        self.inner.outcome_aggregator.send(TrackOutcome {
+            timestamp,
+            scoping,
+            outcome: Outcome::RateLimited(reason_code),
+            event_id: None,
+            remote_addr: None,
+            category: DataCategory::Metrics,
+            quantity: total_batches as u32,
+        });
+    }
+
+    /// Returns `true` if the batches should be rate limited.
+    fn rate_limit_batches(
+        &self,
+        cached_rate_limits: RateLimits,
+        scoping: Scoping,
+        bucket_qty: usize,
+        bucket_partitions: &BTreeMap<Option<u64>, Vec<Bucket>>,
+        max_batch_size_bytes: usize,
+        project_state: Arc<ProjectState>,
+    ) -> bool {
+        let mode = {
+            let usage = match project_state.config.transaction_metrics {
+                Some(ErrorBoundary::Ok(ref c)) => c.usage_metric(),
+                _ => false,
+            };
+            ExtractionMode::from_usage(usage)
+        };
+
+        let quotas = &project_state.config.quotas;
+        let item_scoping = ItemScoping {
+            category: DataCategory::Metrics,
+            scoping: &scoping,
+        };
+
+        // We want to protect the kafka metrics ingestion topics, therefore we ratelimit based on
+        // batches, as the batches are what's sent to kafka.
+        let total_batches: usize = bucket_partitions
+            .values()
+            .map(|buckets| {
+                // Cheap operation because there's no allocations.
+                BucketsView::new(buckets)
+                    .by_size(max_batch_size_bytes)
+                    .count()
+            })
+            .sum();
+
+        // Local check if throughput limit has been exceeded. Used for limiting the amount of redis
+        // calls, as well as extending the throughput limiting to non-processing relays.
+        if cached_rate_limits
+            .check_with_quotas(quotas, item_scoping)
+            .is_limited()
+        {
+            relay_log::info!("dropping {bucket_qty} buckets due to throughput ratelimit");
+            let reason_code = cached_rate_limits
+                .longest()
+                .and_then(|limit| limit.reason_code.clone());
+
+            self.drop_buckets_with_outcomes(
+                reason_code,
+                total_batches,
```

Comment:
What's the difference between `total_batches` and `bucket_qty` (function param), the maximum size of the bucket? Can we use just one of the two? It seems odd to me that sometimes we use one and sometimes the other (see when emitting outcomes and on log messages).

Reply:
`bucket_qty` is the amount of total buckets in the flash, `total_batches` is the amount of batches we are writing upstream (aka. amount of requests to upstream or amount of messages in the kafka topic).